### PR TITLE
ci: add image scan as part of CI

### LINKED ISSUE #415: Enable image scan as part of CI
**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]
- Enable image scan using trivy as part of CI 

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Secrets Store CSI Driver version: (use the image tag):
- Kubernetes version: (use `kubectl version`):

### PR DESCRIPTION
**Pull request by aramase**
**What this PR does / why we need it**:
- Refactors makefile to support multi-os builds. After this PR we should create a follow up to remove `build.sh` used for cloudbuilds.
- Adds target for image-scan

**Which issue(s) this PR fixes** *(optional, using `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when the PR gets merged)*:
Fixes #415

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by aramase** [comment]
/uncc @ritazh @tam7t 

(will tag after ready for review)

File: Makefile [hunk 1]
```diff
@@ -12,138 +12,258 @@
...
+## Tooling Binaries
+## --------------------------------------
 
-# USED for windows CI tests
-.PHONY: e2e-test
-e2e-test: e2e-bootstrap
-	make e2e-azure
+$(CONTROLLER_GEN): $(TOOLS_MOD_DIR)/go.mod $(TOOLS_MOD_DIR)/go.sum $(TOOLS_MOD_DIR)/tools.go ## Build controller-gen from tools folder.
+	cd $(TOOLS_MOD_DIR) && \
+		go build -tags=tools -o $(TOOLS_BIN_DIR)/controller-gen sigs.k8s.io/controller-tools/cmd/controller-gen
```
Comment (ritazh) [inline Makefile:119]: Do we want to pin it to a version?

File: Makefile [hunk 2]
```diff
@@ -12,138 +12,258 @@
...
-	# Download and install Vault
-	vault -v | grep -q v$(VAULT_VERSION) || (curl -LO https://releases.hashicorp.com/vault/$(VAULT_VERSION)/vault_$(VAULT_VERSION)_linux_amd64.zip && unzip vault_$(VAULT_VERSION)_linux_amd64.zip && chmod +x vault && mv vault /usr/local/bin/)
-	# Download and install Azure CLI
+$(GOLANGCI_LINT): ## Build golangci-lint from tools folder.
+	cd $(TOOLS_MOD_DIR) && \
+		go build -tags=tools -o $(TOOLS_BIN_DIR)/golangci-lint github.com/golangci/golangci-lint/cmd/golangci-lint
+
+$(KUSTOMIZE): ## Build kustomize from tools folder.
+	cd $(TOOLS_MOD_DIR) && \
+		go build -tags=tools -o $(TOOLS_BIN_DIR)/kustomize sigs.k8s.io/kustomize/kustomize/v3
```
Comment (ritazh) [inline Makefile:127]: pin it to a version?

File: Makefile (hunk 1 above)
Comment (ritazh) [reply Makefile:119]: nvm i see it in go.mod

File: Makefile (hunk 2 above)
Comment (ritazh) [reply Makefile:127]: nvm i see it in go.mod

**Comment by ritazh** [comment]
cc @sozercan for review

File: Makefile [hunk 3]
```diff
@@ -12,138 +12,258 @@
...
+
+$(KIND): ## Download and install kind
+	kind --version | grep -q $(KIND_VERSION) || (curl -L https://github.com/kubernetes-sigs/kind/releases/download/v$(KIND_VERSION)/kind-linux-amd64 --output kind && chmod +x kind && mv kind /usr/local/bin/)
+
+$(AZURE_CLI): ## Download and install azure cli
 	curl -sL https://aka.ms/InstallAzureCLIDeb | bash
-ifndef TEST_WINDOWS
+
+$(BATS): ## Install bats for running the tests
+	bats --version | grep -q $(BATS_VERSION) || (apt-get -o Acquire::Retries=30 update && apt-get -o Acquire::Retries=30 install -y bats)
```
Comment (sozercan) [inline Makefile:143]: these are different binaries/repos/versions. looks like you are checking for v1.2.1 which is [bats-core](https://github.com/bats-core/bats-core) but I believe apt installed one is [bats](https://github.com/sstephenson/bats) which is the unmaintained older version

File: Makefile [hunk 4]
```diff
@@ -12,138 +12,258 @@
...
+
+$(KUBECTL): ## Install kubectl
 	curl -LO https://storage.googleapis.com/kubernetes-release/release/v$(KUBERNETES_VERSION)/bin/linux/amd64/kubectl && chmod +x ./kubectl && mv kubectl /usr/local/bin/
-	make setup-kind
+
+$(TRIVY): ## Install trivy for image vulnerability scan
+	trivy -v | grep -q $(TRIVY_VERSION) || (curl -sfL https://raw.githubusercontent.com/aquasecurity/trivy/main/contrib/install.sh | sh -s -- -b /usr/local/bin v$(TRIVY_VERSION))
+
+$(BATS): ## Install bats for running the tests
```
Comment (sozercan) [inline Makefile:151]: is this duplicate?

File: Makefile [hunk 5]
```diff
@@ -12,138 +12,258 @@
...
+.PHONY: test-style
+test-style: lint lint-charts
+
+.PHONY: lint
+lint: $(GOLANGCI_LINT)
+	# Setting timeout to 5m as default is 1m
+	$(GOLANGCI_LINT) run --timeout=5m -v
+
+lint-full: $(GOLANGCI_LINT)
+	$(GOLANGCI_LINT) run -v --fast=false
```
Comment (sozercan) [inline Makefile:169]: isn't `fast=false` the default? i think this is same as above.

File: Makefile [hunk 6]
```diff
@@ -12,138 +12,258 @@
...
-		kind load docker-image --name kind e2e/secrets-store-csi:$(IMAGE_VERSION)
+	$(MAKE) container
+	kind load docker-image --name kind $(IMAGE_TAG)
 endif
 
-.PHONY: install-helm
-install-helm:
-	helm version --short | grep -q v3 || (curl https://raw.githubusercontent.com/helm/helm/master/scripts/get-helm-3 | bash)
+.PHONY: e2e-test
+e2e-test: e2e-bootstrap # run test for windows
```
Comment (sozercan) [inline Makefile:262]: is this windows only tests?

File: Makefile [hunk 7]
```diff
@@ -12,138 +12,258 @@
...
+	docker manifest create --amend $(IMAGE_TAG) $(foreach osarch, $(ALL_OS_ARCH), $(IMAGE_TAG)-${osarch})
+	# add "os.version" field to windows images (based on https://github.com/kubernetes/kubernetes/blob/master/build/pause/Makefile)
+	set -x; \
+	registry_prefix=$(shell (echo ${REGISTRY} | grep -Eq ".*[\/\.].*") && echo "" || echo "docker.io/"); \
+	manifest_image_folder=`echo "$${registry_prefix}${IMAGE_TAG}" | sed "s|/|_|g" | sed "s/:/-/"`; \
+	for arch in $(ALL_ARCH.windows); do \
+		for osversion in $(ALL_OSVERSIONS.windows); do \
+			BASEIMAGE=mcr.microsoft.com/windows/nanoserver:$${osversion}; \
+			full_version=`docker manifest inspect $${BASEIMAGE} | jq -r '.manifests[0].platform["os.version"]'`; \
+			sed -i -r "s/(\"os\"\:\"windows\")/\0,\"os.version\":\"$${full_version}\"/" "${HOME}/.docker/manifests/$${manifest_image_folder}/$${manifest_image_folder}-windows-$${osversion}-$${arch}"; \
```
Comment (sozercan) [inline Makefile:230]: what version of Docker runs in actions? can we replace this with `docker manifest annotate`?

File: Makefile [hunk 8]
```diff
@@ -12,138 +12,258 @@
...
+container-windows: docker-buildx-builder
+	docker buildx build --no-cache --output=type=$(OUTPUT_TYPE) --platform="windows/$(ARCH)" --build-arg LDFLAGS=$(LDFLAGS) \
+		--build-arg BASEIMAGE=mcr.microsoft.com/windows/nanoserver:$(OSVERSION) \
+		--build-arg BASEIMAGE_CORE=mcr.microsoft.com/windows/servercore:$(OSVERSION) \
+ 		-t $(IMAGE_TAG)-windows-$(OSVERSION)-$(ARCH) -f docker/windows.Dockerfile .
+
+.PHONY: docker-buildx-builder
+docker-buildx-builder:
+	@if ! docker buildx ls | grep -q container-builder; then\
+		DOCKER_CLI_EXPERIMENTAL=enabled docker buildx create --name container-builder --use --driver-opt image=moby/buildkit:$(BUILDKIT_VERSION);\
```
Comment (sozercan) [inline Makefile:209]: what is the reason for specific `BUILDKIT_VERSION`?

**Review by sozercan — COMMENTED** [review]
looks good, few minor comments

File: Makefile (hunk 6 above)
Comment (aramase) [reply Makefile:262]: that's correct. The target is baked in kubetest: https://github.com/kubernetes/test-infra/blob/master/kubetest/aksengine.go#L1346-L1351

File: Makefile (hunk 8 above)
Comment (aramase) [reply Makefile:209]: The `BUILDKIT_VERSION` was initially introduced when there was a bug that caused windows image builds to fail. We can eventually remove this and rely on the default versions on the host or prow.

File: Makefile (hunk 3 above)
Comment (aramase) [reply Makefile]: Updated it!

File: Makefile (hunk 5 above)
Comment (aramase) [reply Makefile:166]: By default we set it to `true` (https://github.com/kubernetes-sigs/secrets-store-csi-driver/pull/441/files#diff-6179837f7df53a6f05c522b6b7bb566d484d5465d9894fb04910dd08bb40dcc9R23).

File: Makefile (hunk 7 above)
Comment (aramase) [reply Makefile:227]: cloudbuild uses docker version `19.03.5` and `--os-version` isn't supported in that with docker manifest annotate. We can update to using `docker manifest annotate` once the docker version is updated.

**Review by sozercan — APPROVED** [review]
LGTM

**Pull request merged**